src/postprocess/phonon.jl

```diff
@@ -13,18 +13,88 @@ function dynmat_red_to_cart(model::Model, dynamical_matrix)
     #   ⇒ D_cart = lattice⁻ᵀ · D_red · lattice⁻¹.
 
     cart_mat = zero.(dynamical_matrix)
-    for τ in 1:size(cart_mat, 2), η in 1:size(cart_mat, 4)
-        cart_mat[:, η, :, τ] = inv_lattice' * dynamical_matrix[:, η, :, τ] * inv_lattice
+    for s = 1:size(cart_mat, 2), β = 1:size(cart_mat, 4)
+        cart_mat[:, β, :, s] = inv_lattice' * dynamical_matrix[:, β, :, s] * inv_lattice
     end
     cart_mat
 end
 
+# Create a tensor equivalent to a``(n_{\rm atoms}×n_{\rm dim})^2`` diagonal matrix with the
+# atomic masses of the atoms.
+function get_masses(basis::PlaneWaveBasis{T}) where {T}
+    model = basis.model
+    positions = model.positions
+    n_atoms = length(positions)
+
+    atoms_mass = atomic_mass.(model.atoms)
+    any(iszero.(atoms_mass)) && @error "Some elements have unknown masses"
+    masses = zeros(T, 3, n_atoms, 3, n_atoms)
+    for s in eachindex(atoms_mass)
+        masses_s = @view masses[:, s, :, s]
+        masses_s[diagind(masses_s)] .= atoms_mass[s]
+    end
+    masses
+end
+
+@doc raw"""
+Solve the eigenproblem for a dynamical matrix. The phonon frequencies that are returned are
+real numbers in ``{\rm cm}^{-1}``.
+"""
+function phonon_modes(basis::PlaneWaveBasis{T}, dynamical_matrix) where {T}
+    model = basis.model
+    positions = model.positions
+    n_atoms = length(positions)
+
+    mass_mat = reshape(get_masses(basis), 3*n_atoms, 3*n_atoms)
+    cart_mat = reshape(dynmat_red_to_cart(model, dynamical_matrix), 3*n_atoms, 3*n_atoms)
+
+    res = eigen(cart_mat, mass_mat)
+    norm(imag(res.values)) > sqrt(eps(T)) && @warn "Some large imaginary phonon modes"
```

Comment:
careful that people use "imaginary phonon" for something else (omega^2 negative, meaning instabilities)